main/lsp/wrapper.cc

```diff
@@ -1,75 +1,254 @@
 #include "main/lsp/wrapper.h"
 #include "core/errors/namer.h"
+#include "main/lsp/LSPInput.h"
 #include "main/lsp/LSPOutput.h"
+#include "main/lsp/json_types.h"
 #include "main/pipeline/pipeline.h"
 #include "payload/payload.h"
+#include <memory>
 #include <regex>
 
 using namespace std;
 
 namespace sorbet::realmain::lsp {
+namespace {
+void fixGS(core::GlobalState &gs, options::Options &opts) {
+    if (!opts.stripeMode) {
+        // Definitions in multiple locations interact poorly with autoloader this error is enforced
+        // in Stripe code.
+        gs.suppressErrorClass(sorbet::core::errors::Namer::MultipleBehaviorDefs.code);
+    }
+
+    // If we don't tell the errorQueue to ignore flushes, then we won't get diagnostic messages.
+    gs.errorQueue->ignoreFlushes = true;
+
+    // Ensure LSP is enabled.
+    opts.runLSP = true;
+}
+
+enum class FenceState {
```

Comment:
I didn't carefully check that graph of state transitions and how it works with concurrency/implications of it.